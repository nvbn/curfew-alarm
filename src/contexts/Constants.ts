import { createContext } from "react";

import IConstants, { ConstantsDefaultImpl } from "../dependencies/IConstants";

const Constants = createContext<IConstants>(ConstantsDefaultImpl);

export default Constants;
