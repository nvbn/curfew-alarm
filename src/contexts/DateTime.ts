import { createContext } from "react";

// Only to make testing easier
const DateTime = createContext(() => new Date());

export default DateTime;
