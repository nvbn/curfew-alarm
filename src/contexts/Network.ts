import * as ExpoNetwork from "expo-network";
import { createContext } from "react";

import INetwork from "../dependencies/INetwork";

const Network = createContext<INetwork>(ExpoNetwork);

export default Network;
