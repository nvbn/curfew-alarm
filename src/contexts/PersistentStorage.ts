import { createContext } from "react";

import IPersistentStorage, {
  PersistentStorageDefaultImpl,
} from "../dependencies/IPersistentStorage";

const PersistentStorage = createContext<IPersistentStorage>(
  PersistentStorageDefaultImpl,
);

export default PersistentStorage;
