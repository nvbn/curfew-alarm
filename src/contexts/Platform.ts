import { createContext } from "react";

import IPlatform, { PlatformDefaultImpl } from "../dependencies/IPlatform";

const Platform = createContext<IPlatform>(PlatformDefaultImpl);

export default Platform;
