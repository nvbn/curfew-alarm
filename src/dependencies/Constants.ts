import ExpoConstants from "expo-constants";
import { createContext } from "react";

/**
 * Interface that provides access to runtime constants.
 */
export interface IConstants {
  isDevice: boolean;
}

export const ConstantsDefaultImpl: IConstants = ExpoConstants;

export const ConstantsCtx = createContext<IConstants>(ConstantsDefaultImpl);
