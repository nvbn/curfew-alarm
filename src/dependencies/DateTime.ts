import { createContext } from "react";

export type IDateTime = () => Date;

export const DateTimeDefaultImpl: IDateTime = () => new Date();

export const DateTimeCtx = createContext<IDateTime>(DateTimeDefaultImpl);
