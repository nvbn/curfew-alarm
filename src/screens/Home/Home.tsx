import { StackNavigationProp } from "@react-navigation/stack";
import { StatusBar } from "expo-status-bar";
import React, { useEffect, useState } from "react";
import { View } from "react-native";

import Clock from "../../components/Clock";
import useGetDate from "../../hooks/useGetDate";
import useIsAtHome from "../../hooks/useIsAtHome";
import useNotifications from "../../hooks/useNotifications";
import useSettings from "../../hooks/useSettings";
import { isReady } from "../../utils/future";
import { dateToTime } from "../../utils/time";
import styles from "./styles";

type Props = {
  // eslint-disable-next-line
  navigation: StackNavigationProp<{}>;
};

const UPDATE_TIME_INTERVAL = 1000;

/**
 * The home screen of the app.
 */
const Home = ({ navigation }: Props): JSX.Element => {
  useNotifications();

  const [refreshCounter, setRefreshCounter] = useState(0);
  const getDate = useGetDate();
  const [settings] = useSettings([refreshCounter]);
  const isAtHome = useIsAtHome([refreshCounter]);

  useEffect(
    () =>
      navigation.addListener("focus", () => setRefreshCounter((v) => v + 1)),
    [navigation],
  );

  const [now, setNow] = useState(dateToTime(new Date()));
  useEffect(() => {
    const interval = setInterval(
      () => setNow(dateToTime(getDate())),
      UPDATE_TIME_INTERVAL,
    );

    return () => {
      clearInterval(interval);
    };
  }, [setNow, getDate]);

  return (
    <View style={styles.container}>
      {isReady(settings) && (
        <Clock
          curfewStart={settings.curfewStart}
          curfewEnd={settings.curfewEnd}
          minutesToGoHome={settings.minutesToGoHome}
          currentTime={now}
          isAtHome={isReady(isAtHome) ? isAtHome : true}
        />
      )}
      <StatusBar style="auto" />
    </View>
  );
};

export default Home;
