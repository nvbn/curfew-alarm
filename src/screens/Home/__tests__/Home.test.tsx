/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-empty-function */
import { render } from "@testing-library/react-native";
import React from "react";

import Constants from "../../../contexts/Constants";
import DateTime from "../../../contexts/DateTime";
import Network from "../../../contexts/Network";
import PersistentStorage from "../../../contexts/PersistentStorage";
import Platform from "../../../contexts/Platform";
import { makeConstantsAsOnDevice } from "../../../fakes/Constants";
import { makeNetworkAsOnWifi } from "../../../fakes/Network";
import { makePersistentStorageWithDataAndBehavior } from "../../../fakes/PersistentStorage";
import { makePlatformAndroid, makePlatformIOS } from "../../../fakes/Plaftorm";
import Home from "../Home";

describe("<Home />", () => {
  for (const platform of [makePlatformAndroid(), makePlatformIOS()]) {
    test(`can be rendered on ${platform}`, () => {
      const rendered = render(
        <DateTime.Provider value={() => new Date("2021-01-01T12:15:00 GMT")}>
          <PersistentStorage.Provider
            value={makePersistentStorageWithDataAndBehavior()}
          >
            <Network.Provider value={makeNetworkAsOnWifi()}>
              <Platform.Provider value={platform}>
                <Constants.Provider value={makeConstantsAsOnDevice()}>
                  <Home navigation={{ addListener: () => () => {} } as any} />
                </Constants.Provider>
              </Platform.Provider>
            </Network.Provider>
          </PersistentStorage.Provider>
        </DateTime.Provider>,
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });
  }
});
