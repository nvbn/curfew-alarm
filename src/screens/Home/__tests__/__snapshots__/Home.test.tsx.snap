// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> can be rendered on [object Object] 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
/>
`;

exports[`<Home /> can be rendered on [object Object] 2`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
/>
`;
